### DISCUSSION TITLE
Unable to resolve service for type 'Microsoft.AspNetCore.SignalR.Hub`1[Ixxxx]' while attempting to activate 'xxxxx'

### DISCUSSION THREAD

**Discussion by ViacheslavUKR** [dotnet/aspnetcore#25166] [General] [answered]
I define my typed SignalR Hub service as
```
            services.AddSignalR()
                    .AddHubOptions<OutputMessages>(options =>
                    {
                        options.EnableDetailedErrors = true;
                    });
           app.UseEndpoints(endpoints =>
                   {
                       endpoints.MapHub<OutputMessages>("/OutputMessages", options =>
                       {
                             options.Transports =
                             HttpTransportType.WebSockets |
                             HttpTransportType.LongPolling;
                       });
                   });
```
Then define interface IOutputMessage and realization:

```
    public class OutputMessages : Hub<IOutputMessages>
    {
        public OutputMessages(Hub<IOutputMessages> hubContext, ILogger<OutputMessages> logger, ApplicationDbContext dbContext)
        {
```
Than inject this service to comtroller
```
    public class ApplicationAPIController : ControllerBase
    {
        public ApplicationAPIController(ILogger<ApplicationAPIController> logger, ApplicationDbContext dbContext, IConfiguration Configuration, CoreObjectDumper.CoreObjectDumper dump, IHubContext<OutputMessages, IOutputMessages> _outputMessages)
...
              outputMessages.Clients.All.AdminMessages2(sessionData.Result.UserId, "Used Stop", false);
```
After I try to connect to my SignalR hub I receive 

```
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint '/OutputMessages'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 87.8426ms 101
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 POST http://localhost:5000/OutputMessages/negotiate?negotiateVersion=1  0
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint '/OutputMessages/negotiate'
dbug: Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager[1]
      New connection tphQJH1K2qR9XAtDyRbH5w created.
dbug: Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher[10]
      Sending negotiation response.
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint '/OutputMessages/negotiate'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 19.5426ms 200 application/json
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/OutputMessages?id=JQSyzJUvPHGEcquo1UwfrA
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint '/OutputMessages'
dbug: Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher[4]
      Establishing new connection.
dbug: Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport[1]
      Socket opened using Sub-Protocol: '(null)'.
dbug: Microsoft.AspNetCore.SignalR.HubConnectionHandler[5]
      OnConnectedAsync started.
dbug: Microsoft.AspNetCore.SignalR.Internal.DefaultHubProtocolResolver[2]
      Found protocol implementation for requested protocol: json.
dbug: Microsoft.AspNetCore.SignalR.HubConnectionContext[1]
      Completed connection handshake. Using HubProtocol 'json'.
fail: Microsoft.AspNetCore.SignalR.HubConnectionHandler[1]
      Error when dispatching 'OnConnectedAsync' on hub.
System.InvalidOperationException: Unable to resolve service for type 'Microsoft.AspNetCore.SignalR.Hub`1[SignalR.IOutputMessages]' while attempting to activate 'SignalR.OutputMessages'.
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(IServiceProvider sp, Type type, Type requiredBy, Boolean isDefaultParameterRequired)
   at lambda_method(Closure , IServiceProvider , Object[] )
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubActivator`1.Create()
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.OnConnectedAsync(HubConnectionContext connection)
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.OnConnectedAsync(HubConnectionContext connection)
   at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.RunHubAsync(HubConnectionContext connection)
dbug: Microsoft.AspNetCore.SignalR.HubConnectionHandler[6]
      OnConnectedAsync ending.
dbug: Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport[4]
      Waiting for the application to finish sending data.
dbug: Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport[2]
      Socket closed.
dbug: Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager[2]
      Removing connection JQSyzJUvPHGEcquo1UwfrA from the list of connections.
```

What I doing wrong?

**Comment by BrennanConroy** [comment, marked as answer]
Why is `Hub<IOutputMessages> hubContext` in the constructor of your Hub? That's what is causing issues.

**Reply by ViacheslavUKR** [reply]
my pervious version is     
public class OutputMessages : Hub&lt;OutputMessages&gt;
as a result I must inject service to controller not as IHubContext<OutputMessages, IOutputMessages> _outputMessages
but as IHubContext &lt;OutputMessages&gt; _outputMessages
as a result I don't have access to my typed method of Hub, namely instead `outputMessages.Clients.All.AdminMessages2(sessionData.Result.UserId, "Used Stop", false);`
i must use 
outputMessages.Clients.All.SendAsync("AdminMessages2", sessionData.Result.UserId, "Used Stop", false);
and moreover finally result is not changed. I receive the same error message "Unable to resolve service...:.

**Reply by BrennanConroy** [reply]
Take a look at https://docs.microsoft.com/en-us/aspnet/core/signalr/hubcontext?view=aspnetcore-3.1#inject-a-strongly-typed-hubcontext
You should be able to use a strongly-typed HubContext, but again, you shouldn't have `Hub<T>` in your constructor.

**Reply by ViacheslavUKR** [reply]
Thank you. Unfortunately your link not contains any significant information, nothing about definition Hub, nothing about definition Service, nothing a workable project. This link contains only way to inject SignalR Hub to controller. Yes, I inject exactly i the same way (in old version of my code - in new version as public ChatController(IHubContext<ChatHub> chatHubContext) and after that I receive an error
![Blazor-T232020_202604](https://user-images.githubusercontent.com/20413407/90986642-12654e80-e58d-11ea-863f-dbb739112c87.png)

```

**Reply by BrennanConroy** [reply]
Change your hubs constructor
`public OutputMessages(IHubContext<OutputMessages, IOutputMessages> hubContext, ILogger<OutputMessages> logger, ApplicationDbContext dbContext)`

**Reply by ViacheslavUKR** [reply]
This changing need another definition of Hub maybe, directly this is impossible
![Blazor-T232020_222103](https://user-images.githubusercontent.com/20413407/90986930-71c45e00-e58f-11ea-9c64-5b138cf9e07d.png)

**Reply by BrennanConroy** [reply]
That's because you changed your hub to be `public class OutputMessages : Hub<OutputMessages>` instead of `public class OutputMessages : Hub<IOutputMessages>`, change it back.

**Reply by ViacheslavUKR** [reply]
I return to first version of my code (with interface in Hub constructor), but error message not changing.

![Blazor-T242020_212715](https://user-images.githubusercontent.com/20413407/91082296-2a9ea180-e651-11ea-833b-1dc947f156a4.png)

This is my Hub Interface
```
    public interface IOutputMessages
    {
        Task AdminMessages(string message, bool? consoleOnly);
        Task AdminMessages2(string userId, string message, bool consoleOnly);
        Task CrossServerNotice(string discordId, string server, string message);
        Task NewDepositNotice(string discordId, string coin, double amount, int confirms, bool complete);
        Task NewMasternodePayoutNotice(string discordId, string pool, double costpershare, int shares, double payoutpershare, double newBalance);
        Task NewSharePurchase(string discordId, string pool, double costpershare, int shares, double total, int newcount);
        Task NewShareSale(string discordId, string pool, double costpershare, int shares, double total);
        Task OnConnectedAsync();
        Task OnDisconnectedAsync(Exception exception);
        Task SendUserMessage(string discordId, Guid containerId, string message);
        Task SendUserMessageNoContainer(string discordId, string message);
        Task UserPayoutNotice(string discordId, string server, string message);
    }
```
This is my Hub Code

```
 public class OutputMessages : Hub<IOutputMessages>
    {
        public static readonly Dictionary<string, string> ConnectedUserList = new Dictionary<string, string>(1024);
        private readonly Hub<IOutputMessages> _hubContext;
        private readonly ApplicationDbContext _db;
        private readonly ILogger<OutputMessages> _logger;

        public OutputMessages(Hub<IOutputMessages> hubContext, ILogger<OutputMessages> logger, ApplicationDbContext dbContext)
        {
            _hubContext = hubContext;
            _db = dbContext;
            _logger = logger;
        }
       ...
        public async Task AdminMessages2(string userId, string message, bool consoleOnly)
        {
            string uname = await Username(userId);
            await _hubContext.Clients.All.AdminMessages2(userId, "API: " + uname + " " + message, consoleOnly);
        }

        [ResponseCache(VaryByHeader = "id", Duration = 36000, Location = ResponseCacheLocation.Any)]
        public async Task<string> Username(string id)
        {
            ApplicationUser user = null;
            await Task.Run(() => user = _db.Users.Find(id));
            return user.UserName;
        }

        public override async Task OnConnectedAsync()
        {
            string JWT = Context.GetHttpContext().Request.Headers["Authorization"];
            string ValidUserID = UserHelpers.ValidateJWT(JWT);
            if (ValidUserID != null)
            {
                _logger.LogInformation($"User {ValidUserID} connected to {this.GetType().Name} hub");
                if (ConnectedUserList.Any(x => x.Key == ValidUserID))
                {
                    ConnectedUserList.Remove(ValidUserID);
                }
                ConnectedUserList.Add(ValidUserID, Context.ConnectionId);
            }
            await base.OnConnectedAsync();
        }

        public override async Task OnDisconnectedAsync(System.Exception exception)
        {
            var JWT = Context.GetHttpContext().Request.Headers["Authorization"];
            string ValidUserID = UserHelpers.ValidateJWT(JWT);
            if (ValidUserID != null)
            {
                _logger.LogInformation($"User {ValidUserID} disconnected from {this.GetType().Name} hub");
                if (ConnectedUserList.Any(x => x.Key == ValidUserID))
                {
                    ConnectedUserList.Remove(ValidUserID);
                }
                ConnectedUserList.Add(ValidUserID, Context.ConnectionId);
            }
            await base.OnDisconnectedAsync(exception);
        }
    }
```
 In this case maybe injection in control doing correctly, I injecting SignalR service to controller as 

`        public ApplicationAPIController(ILogger<ApplicationAPIController> logger, ApplicationDbContext dbContext, IConfiguration Configuration, CoreObjectDumper.CoreObjectDumper dump, IHubContext<OutputMessages, IOutputMessages> _outputMessages)`

But code don't working and I don't understand what's happening. Sorry.

**Reply by BrennanConroy** [reply]
You still haven't changed your hubs constructor https://github.com/dotnet/aspnetcore/discussions/25166#discussioncomment-55038

**Reply by ViacheslavUKR** [reply]
Directly it's impossible. Maybe need something more...
![Blazor-T242020_232109](https://user-images.githubusercontent.com/20413407/91092430-b23fdc80-e660-11ea-8f19-a58015aff342.png)

**Reply by BrennanConroy** [reply]
Why did you change your hub definition back to the wrong version?

Use `IHubContext<OutputMessages, IOutputMessages>` in the constructor and keep the hub defined as `public class OutputMessages : Hub<IOutputMessages>`

**Reply by ViacheslavUKR** [reply]
![Blazor-T242020_232109](https://user-images.githubusercontent.com/20413407/91098653-89244980-e66a-11ea-9e5b-04958f0390c4.png)

it's impossible, sorry. If I type this line

`public OutputMessages(IHubContext<OutputMessages, IOutputMessages> hubContext, ILogger<OutputMessages> logger, ApplicationDbContext dbContext)`

I have received compilation error

`Error	CS0311	The type 'OutputMessages' cannot be used as type parameter 'THub' in the generic type or method 'IHubContext<THub, T>'. There is no implicit reference conversion from 'OutputMessages' to 'Microsoft.AspNetCore.SignalR.Hub<IOutputMessages>'.
`

I suspect something wrong in Hub class or in DJ processing of SignalR Hub. Because there are nothing anyone full workable example in whole inet with strong typed Hub. Only small fragments of non accessible projects, even in this repository is exist non workable example with strong type Hub - strong type class is defined, but not using in project.

**Reply by BrennanConroy** [reply]
You are still defining your hub wrong. Please just copy paste this as your class definition: `public class OutputMessages : Hub<IOutputMessages>`

**Reply by ViacheslavUKR** [reply]
yes, thank you for your efforts. Finally connection is ON. 
![Blazor-T252020_021504](https://user-images.githubusercontent.com/20413407/91105674-5d5c9000-e679-11ea-9922-a19b7a409dbf.png)

Message still not received for client, but maybe it's another problem with Blazor frontend or... I don't know what happens, I need more time. 
But DJ container and SignalR Hub constructor now working correctly.  And injection Hub to controller is working.
Thank you.
